File: ruby/bracket-push/bracket_push.rb
# frozen_string_literal: true

module Brackets
  OPENING_BRACKETS = "([{"
  private_constant :OPENING_BRACKETS
  CLOSING_BRACKETS = ")]}"
  private_constant :CLOSING_BRACKETS
  WHITESPACE = /\s/.freeze
  private_constant :WHITESPACE

  module_function

  def paired?(string)
    string = remove_white_space(string)
    return true if string.empty?

    catch(:halt) do
      string
        .each_char
        .with_object([], &method(:check_pair))
        .empty?
    end
  end

  def remove_white_space(string)
    string.gsub(WHITESPACE, "")
  end
  private_class_method :remove_white_space

  def check_pair(char, acc)
    throw(:halt, false) if
      closing_bracket?(char) && no_matching_opening_bracket?(acc, char)

    acc.push(char) if opening_bracket?(char)
  end
  private_class_method :check_pair

  def closing_bracket?(char)
    CLOSING_BRACKETS.include?(char)
  end
  private_class_method :closing_bracket?

  def no_matching_opening_bracket?(acc, char)
    acc.pop != char.tr(CLOSING_BRACKETS, OPENING_BRACKETS)
  end

  def opening_bracket?(char)
    OPENING_BRACKETS.include?(char)
  end
  private_class_method :opening_bracket?
end

File: elixir/bracket-push/lib/bracket_push.ex
defmodule BracketPush do
  @opening_brackets '([{'
  @closing_brackets ')]}'

  defguardp opening_bracket?(char) when char in @opening_brackets
  defguardp closing_bracket?(char) when char in @closing_brackets

  @doc """
  Checks that all the brackets and braces in the string are matched correctly, and nested correctly
  """
  @spec check_brackets(String.t()) :: boolean
  def check_brackets(str) do
    str
    |> String.to_charlist()
    |> Enum.reduce([], &check_pair/2)
    |> Enum.empty?()
  catch
    :halt ->
      false
  end

  defp check_pair(char, acc) when opening_bracket?(char), do: [char | acc]

  defp check_pair(char, acc) when closing_bracket?(char) do
    if matching_opening_bracket?(char, acc) do
      tl(acc)
    else
      throw(:halt)
    end
  end

  defp check_pair(_char, acc), do: acc

  defp matching_opening_bracket?(_char, []), do: false

  defp matching_opening_bracket?(char, acc) do
    opening_bracket =
      @closing_brackets
      |> Enum.find_index(&(&1 == char))
      |> (fn index -> Enum.at(@opening_brackets, index) end).()

    hd(acc) == opening_bracket
  end
end
